flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/sink/abilities/SupportsPartitioning.java

```diff
@@ -0,0 +1,109 @@
+ * of the partition keys (in the example: first by region then by month) is defined by the catalog table. A
+ * list of partitions could be:
+ * <pre>
+ *   List(
+ *     ['region'='europe', 'month'='2020-01'],
+ *     ['region'='europe', 'month'='2020-02'],
+ *     ['region'='asia', 'month'='2020-01'],
+ *     ['region'='asia', 'month'='2020-02']
+ *   )
+ * </pre>
+ *
+ * <p>Given the following partitioned table:
+ * <pre>{@code
+ *   CREATE TABLE t (a INT, b STRING, c DOUBLE, region STRING, month STRING) PARTITION BY (region, month);
+ * }</pre>
+ *
+ * <p>We can insert data into <i>static table partitions</i> using the {@code INSERT INTO ... PARTITION} syntax:
+ * <pre>{@code
+ *     INSERT INTO t PARTITION (region='europe', month='2020-01') SELECT a, b, c FROM my_view;
+ * }</pre>
+ *
+ * <p>If all partition keys get a value assigned in the {@code PARTITION} clause, the operation is considered
+ * as an "insertion into a static partition". In the above example, the query result should be written
+ * into the static partition {@code region='europe', month='2020-01'} which will be passed by the planner
+ * into {@link #applyStaticPartition(Map)}.
+ *
+ * <p>Alternatively, we can insert data into <i>dynamic table partitions</i> using the SQL syntax:
+ * <pre>{@code
+ *     INSERT INTO t PARTITION (region='europe') SELECT a, b, c, month FROM another_view;
+ * }</pre>
+ *
+ * <p>If only a subset of all partition keys (a prefix part) get a value assigned in the {@code PARTITION}
+ * clause, the operation is considered as an "insertion into a dynamic partition". In the above example,
+ * the static partition part is {@code region='europe'} which will be passed by the planner into
+ * {@link #applyStaticPartition(Map)}. The remaining values for partition keys should be obtained from
+ * each individual record by the sink during runtime. In the example, the {@code month} field is the dynamic
+ * partition key.
+ */
+@PublicEvolving
+public interface SupportsPartitioning {
+
+	/**
+	 * Provides the static part of a partition.
+	 *
+	 * <p>A single partition maps each partition key to a partition value. Depending on the user-defined
+	 * statement, the partition might not include all partition keys.
+	 *
+	 * <p>See the documentation of {@link SupportsPartitioning} for more information.
+	 *
+	 * @param partition user-defined (possibly partial) static partition
+	 */
+	void applyStaticPartition(Map<String, String> partition);
+
+	/**
+	 * Returns whether data needs to be grouped by partition before it is consumed by the sink. By default,
+	 * this is not required from the runtime and records arrive in arbitrary partition order.
+	 *
+	 * <p>If this method returns true, the sink can expect that all records will be grouped by the partition
+	 * keys before consumed by the sink. In other words: The sink will receive all elements of one
+	 * partition and then all elements of another partition. Elements of different partitions will not
```

Comment:
`The sink will receive all elements of one partition and then all elements of another partition.` -> I don't necessarily understand this statement. Does it mean that a single parallel instance can receive elements of different partitions? How will it know when a partition ends? Does it check the values of the partition fields?

Reply:
Yes. It will create a new writer if partition value changed. See `GroupedPartitionWriter#write`.

Reply:
Does it mean it makes sense only for bounded data?
Isn't it true that: 
bounded sink -> supportsGrouping=true
unbounded sink -> supportsGrouping=false

Reply:
I think window aggregation can also supportsGrouping in the future, if the window value is the partition. e.g. 

```
INSERT INTO hive_table
SELECT TUMBLE_START(ts, INTERVAL '1' DAY) as day, key1, key2, SUM(a), COUNT(*)
FROM T
GROUP BY TUMBLE_START(ts, INTERVAL '1' DAY), key1, key2
```

Where `day` is the partition key.

Reply:
Ok I get it.

Then last suggestion. Of course feel free to ignore it. Maybe it would be better to change it to two separate methods?

```
Set<OutputMode> supportsOutputMode();

void applyOutputMode(OutputMode outputMode);

enum OutputMode {
   GROUPED,
   NON-GROUPED
}
```

Reply:
I'm fine with this change. Maybe we can come up with a better name for `OutputMode`, it sounds very generic. What do you think about this @JingsongLi @lirui-apache ?

Reply:
I don't see a big benefit of separating the methods. `requiresPartitionGrouping` is basically similar to `ChangelogMode getChangelogMode(ChangelogMode requestedMode)`.

Reply:
I don't have a strong opinion here. Perhaps it's fine to go with `requiresPartitionGrouping`, as long as we have good documentation about the `supportsGrouping` parameter.

Reply:
In FLIP-63, we have a lot of discussions, and I don't have a strong tendency.

Reply:
The benefits I see in the approach I suggested are as follows:
* The planner makes the decision which mode is used based on the information provided by the source. In the current interface it's the source that makes that decision. The logic must be reimplemented in every source separately (I know this logic is rather easy, nevertheless)
* It's more extendible. It's easier to add new modes in the future. I understand there is a low chance there will be new modes.
* It's more aligned with other interfaces which have `applyXXX` methods. In the current approach the `requiresPartitionGrouping` has at least two responsibilities. 1) applies a mode (derived from the input parameter),  2) plus communicates the chosen mode to the planner.

Nevertheless I am really not strong on this and I am also ok with the current interface, as it is rather a niche use case that probably applies to the filesystem connector exclusively.